File: binds/go.v1/src/webserver.go
package main

import (
	"fmt"
	"math/rand"
	"time"

	"gopkg.in/inconshreveable/log15.v2"
	"gopkg.in/project-iris/iris-go.v1"
)

func init() {
	rand.Seed(time.Now().UnixNano())
	iris.Log.SetHandler(log15.DiscardHandler())
}

type webserver struct {
	id int
}

func (w *webserver) Init(conn *iris.Connection) error { return nil }
func (w *webserver) HandleBroadcast(msg []byte)       { panic("Not implemented!") }
func (w *webserver) HandleTunnel(tunn *iris.Tunnel)   { panic("Not implemented!") }
func (w *webserver) HandleDrop(reason error)          { panic("Not implemented!") }

// START OMIT

// Format each request a bit and return as the reply
func (w *webserver) HandleRequest(request []byte) ([]byte, error) { // HLreq
	return []byte(fmt.Sprint("go-www-", w.id, ": ", string(request))), nil
}

func main() {
	// Register a webserver micro-service into the network
	service, err := iris.Register(55555, "webserver", &webserver{id: rand.Intn(100)}, nil) // HLreq
	if err != nil {
		fmt.Println("Failed to register micro-service:", err); return
	}
	defer service.Unregister() // HLreq

	fmt.Println("Waiting for inbound requests...")
	time.Sleep(60 * time.Second)
}

// END OMIT

File: binds/erlang.v1/src/webserver.erl
-module(webserver).
-export([main/0]).

-behaviour(iris_server).
-export([init/2, handle_broadcast/2, handle_request/3, handle_tunnel/2,
    handle_drop/2, terminate/2]).

% Implement the startup and cleanup methods of iris_server.
init(_Conn, Id)            -> {ok, Id}.
terminate(_Reason, _State) -> ok.

%% START OMIT
%% Format each request a bit and return as the reply
handle_request(Request, _From, Id) -> // HLreq
	Response = <<Id/binary, ": ", Request/binary>>,
	{reply, {ok, Response}, Id}.

main() ->
    % Generate a random ID for the web server
    random:seed(erlang:now()),
    Id = list_to_binary(io_lib:format("erl-www-~p", [random:uniform(100)])),

    % Register a new webserver into the Iris network
    {ok, Server} = iris_server:start(55555, "webserver", ?MODULE, Id), // HLreq

    % Serve a while, then quit
    io:format("Waiting for requests..."),
    timer:sleep(100 * 1000),

    ok = iris_server:stop(Server).
%% END OMIT

% Remaining callbacks methods, not used in this demo
handle_broadcast(_Message, State) -> {stop, not_implemented, State}.
handle_tunnel(_Tunnel, State)     -> {stop, not_implemented, State}.
handle_drop(_Reason, State)       -> {stop, not_implemented, State}.
